### ISSUE TITLE
tokio::time::sleep panics

### ISSUE THREAD

**Issue by jon-chuang** [oxalica/async-ffi#14]
My Tokio runtime:
```rust
let rt = tokio::runtime::Builder::new_multi_thread()
    .worker_threads(1)
    .enable_all()
    .build()
    .unwrap();
```
I even tried enabling `time` and `net` features.

async:
```rust
rt.spawn(
    async move {
        ffi_fn().await;
    }
}
```

ffi_fn:
```rust
pub extern "C" fn ffi_fn() -> FfiFuture<()>
    async {
        tokio::time::sleep(std::time::Duration::from_millis(500)).await;
    }.into_ffi()
}
```

**Comment by jon-chuang** [comment]
This is the specific error:
https://github.com/tokio-rs/tokio/blob/ed187ddfb8d99682ff104832c798022a69bd7016/tokio/src/runtime/context.rs#L52
```rust
pub(crate) fn time_handle() -> crate::runtime::driver::TimeHandle {
        match CONTEXT.try_with(|ctx| {
            let ctx = ctx.borrow();
            ctx.as_ref().expect(crate::util::error::CONTEXT_MISSING_ERROR).time_handle.clone()
        }) {
            Ok(time_handle) => time_handle,
            Err(_) => panic!("{}", crate::util::error::THREAD_LOCAL_DESTROYED_ERROR),
        }
    }
```

This is due to trying to access this: 
```rust
thread_local! {
    static CONTEXT: RefCell<Option<Handle>> = RefCell::new(None)
}
```

**Comment by jon-chuang** [comment]
There is actually a solution I know of to solve this sort of gnarly issue with thread local.

That is to compile and link tokio as a shared library dependency. That way, the thread-local var will be initialized equally by all parties.

**Comment by jon-chuang** [comment]
I've basically solved this: https://github.com/tokio-rs/tokio/pull/4493

Let me know if you'd like me to try to upstream a macro.

**Comment by oxalica** [comment]
The issue is caused by `tokio` using specific thread local variables to talk with the runtime when calling leaf async functions (`sleep`, etc). Thus nearly all of its functions require to be run inside an initialized `tokio` runtime, and not generally compatible with just `Future` interface. That's also the reason why `tokio` libraries can only be run under `tokio` runtime. If I remember correctly, `async-std` functions are not exclusive and are compatible with any runtime.

If you do want to use `tokio` runtime, or being runtime-agnostic, I'd suggest to export necessary leaf async functions (`sleep`, `write`, `send`, etc) in host side. For client side, don't include any async runtime, and just use these exported async basics to implement your plugin functions. This can also avoid initializing the runtime twice in client side.

**Comment by jon-chuang** [comment]
That's not feasible if I were using one of the libraries built on top of tokio like Tower, Hyper etc. 

I'd like to be able to plug and play. I am now investigating how feasible this is.

Note that I am not, as you assume, instantiating a new runtime in the client side. Rather, I am merely passing a handle to the runtime from the host.

However, thanks for the tips about async-std. 

I have to do a lot more testing now to see what situations my solution works in and in which ones it breaks.

**Comment by oxalica** [comment]
> Note that I am not, as you assume, instantiating a new runtime in the client side. 

**TL;DR**: Blame `tokio`. Try other runtimes to work out-of-box.

Longer version: `tokio` uses hidden shared statics besides the standard `Future`/`Waker` interface, while this crate can only handle the latter one. So `tokio` cannot be used crossing FFI boundary without starting a new runtime. It's an issue of `tokio` itself. Other runtimes like `async-std`, `futures` or your handwriting one, would work without any issue.

> Rather, I am merely passing a handle to the runtime from the host. 

Sorry but you cannot. Because the host and client side are compiled in two batch and cannot co-work. Rust uses static linking now.

Detail version:
1. `tokio` from the host and the client are different and cannot communicate through internal static or thread locals, since they are linked in two batches. So your `sleep` doesn't even know it's inside a `tokio` runtime.
2. Due to its exclusive implementation, `tokio`'s reactor (`sleep`, `write`, `send`, etc), must be called in the same `tokio` runtime, or it will panic.
3. This crate wraps the standard `Future`/`Waker` interfaces. So `Future`s can correctly communicate through it. But we cannot wrap other hidden ways.
3. Other runtime like `async-std`, `futures` are runtime-agnostic because they only use the standard interfaces to communicate with the runtime. If you use their reactor, like `sleep`, they work under any runtime, no matter it's `tokio`, `async-std`, or your handwriting one. (They also don't need to auto-start a new runtime, unless you call `spawn` in client side.)

So my suggestions are,
1. Use non-`tokio` async basics in **client side**. So they always work regardless of the runtime of host side. (See [example](https://github.com/oxalica/async-ffi/tree/master/link_tests), I use `async-std`'s `sleep` in `tokio` runtime. But you cannot do the reverse.)
2. If you stick on `tokio`, you could either start a new `tokio` runtime and carefully write your C-exported functions to not directly using any `tokio` functions. Or,
   > ...export necessary leaf async functions (sleep, write, send, etc) in host side. For client side, don't include any async runtime, and just use these exported async basics to implement your plugin functions. This can also avoid initializing the runtime twice in client side.

**Comment by jon-chuang** [comment]
@oxalica I've managed to solve the problems you described more or less entirely.

I register some callbacks which initialize the thread_local vars from the Handle of the outer callback. 

I will turn this into a crate, `tokio-dynlib`, unless you are interested in also hosting this here? (I think the scope is related but somewhat separate)
